plugin-server/src/main/ingestion-queues/session-recording/session-recordings-consumer-v3.ts

```diff
@@ -0,0 +1,450 @@
+                })
+                .then(async () => {
+                    // If the SessionManager is done (flushed and with no more queued events) then we remove it to free up memory
+                    if (await sessionManager.isEmpty()) {
+                        await this.destroySession(key, sessionManager)
+                    }
+                })
+            promises.push(flushPromise)
+        }
+        await Promise.allSettled(promises)
+        gaugeSessionsHandled.set(Object.keys(this.sessions).length)
+    }
+
+    private async syncSessionsWithDisk(): Promise<void> {
+        // As we may get assigned and reassigned partitions, we want to make sure that we have all sessions loaded into memory
+        await Promise.all(
+            this.assignedTopicPartitions.map(async ({ partition }) => {
+                const keys = await readdir(path.join(this.rootDir, `${partition}`)).catch(() => {
+                    // This happens if there are no files on disk for that partition yet
+                    return []
+                })
+
+                // TODO: Below regex is a little crude. We should fix it
+                await Promise.all(
+                    keys
+                        .filter((x) => /\d+__[a-zA-Z0-9\-]+/.test(x))
+                        .map(async (key) => {
+                            // TODO: Ensure sessionId can only be a uuid
+                            const [teamId, sessionId] = key.split('__')
+
+                            if (!this.sessions[key]) {
+                                this.sessions[key] = await SessionManagerV3.create(this.config, this.objectStorage.s3, {
+                                    teamId: parseInt(teamId),
+                                    sessionId,
+                                    dir: this.dirForSession(partition, parseInt(teamId), sessionId),
+                                    partition,
+                                })
+                            }
+                        })
+                )
+            })
+        )
+    }
+
+    private async destroySession(key: string, sessionManager: SessionManagerV3): Promise<void> {
+        delete this.sessions[key]
+        await sessionManager.stop()
+    }
+
+    private setupHttpRoutes() {
+        expressApp.get('/api/projects/:projectId/session_recordings/:sessionId', async (req, res) => {
+            // TODO: Sanitize the projectId and sessionId as we are checking the filesystem
+
+            // validate that projectId is a number and sessionId is UUID like
+            const projectId = parseInt(req.params.projectId)
+            if (isNaN(projectId)) {
+                res.sendStatus(404)
+                return
+            }
+
+            const sessionId = req.params.sessionId
+            if (!/^[0-9a-f-]+$/.test(sessionId)) {
+                res.sendStatus(404)
+                return
+            }
```

Comment:
we should count or log these - at least at first we'll need to be sure things are clipping together as expected...

maybe express has some middleware that will do this automagically

Reply:
Count or log what exactly? The requests?

Reply:
yeah 12 requests, 2 failed because project id was "undefined", 3 failed bcause session id was "bobby tables"

only really to get us over the hump of "first few times we try it in production and it doesn't work do we know why"